[FEATURE] Allow disabling of "inline style" and "style block" parsing

erikhansen opened the pull request:
This pull request implements the functionality described in task #155.

Please review how I edited the `emogrify()` method to disable "inline style" parsing to ensure it was done properly.

I updated the readme and added unit tests to cover the new options that were added.

Once you merge in this pull request, would you mind tagging a 0.1.1 release?

erikhansen commented:
@oliverklee - The Travis build is failing on this PR. However upon reviewing the errors that PhpCodeSniffer is finding, they are all related to existing code, not code that I've added (except for one issue that I fixed in 2c4fdc5). For example:

* Expected 0 spaces after opening bracket; newline found https://github.com/jjriv/emogrifier/blob/master/Classes/Emogrifier.php#L270 
* Expected 1 space after FUNCTION keyword; 0 found https://github.com/jjriv/emogrifier/blob/master/Classes/Emogrifier.php#L532
* Line indented incorrectly; expected at least 16 spaces, 12 found https://github.com/jjriv/emogrifier/blob/master/Tests/Unit/EmogrifierTest.php#L376

Let me know how you want me to handle the errors.

oliverklee commented:
I've fixed the PHP_CodeSniffer warnings on Travis. Could you please rebase and re-push? Thanks!

Could you also please squash your commits into one and change the [TASK] into [FEATURE]?

oliverklee commented:
(I'll have a look at the code after the re-push.)

erikhansen commented:
@oliverklee - Thanks for fixing the PHP_CodeSniffer warnings. I've rebased my commits on to master and have squashed them into a single commit. Changed the commit tag to [FEATURE], although I left the "task/" label in the tag name, as I wasn't sure if I could change it.

oliverklee commented:
In the commit message:

* Could you please keep the lines in the description under 72 characters?
* Could you please move the "closes" line out of the list to its own line using the format "Fixes #..."?

http://chris.beams.io/posts/git-commit/

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -190,25 +210,32 @@ public function emogrify()
-                        $normalizedOriginalStyle
+                // If inline styles parsing is enabled, normalize the "style" attribute and save it
+                if ($this->isInlineStylesParsingEnabled) {
+                    $normalizedOriginalStyle = preg_replace_callback(
```
While you're changing is anyway, could you please move the code block within the if to a separate private method?

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -277,14 +307,16 @@ function (array $m) {
-            );
+        // If enabled, iterate through the nodes that contained inline styles in the original HTML
+        if ($this->isInlineStylesParsingEnabled) {
+            foreach ($this->styleAttributesForNodes as $nodePath => $styleAttributesForNode) {
```
While you're changing this part anyway, could you please move it to a separate private method?

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -316,6 +348,26 @@ function (array $m) {
     }
 
     /**
+     * Disable the parsing of inline styles
```
Third person, please: DisableS

oliverklee replied:
And please add the period at the end of the sentence.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -316,6 +348,26 @@ function (array $m) {
+     * Disable the parsing of inline styles
+     *
+     * @see $isInlineStylesParsingEnabled
+     */
```
Please add the @return void.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -316,6 +348,26 @@ function (array $m) {
+     * Disables the parsing of <style> blocks
+     *
+     * @see $isStyleBlocksParsingEnabled
+     */
```
Please add the @return void.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -316,6 +348,26 @@ function (array $m) {
+    }
+
+    /**
+     * Disables the parsing of <style> blocks
```
Please add the period at the end of the sentence.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -316,6 +348,26 @@ function (array $m) {
     /**
+     * Disable the parsing of inline styles
+     *
+     * @see $isInlineStylesParsingEnabled
```
I don't think we need this @see as the name of the field is not relevant to the external API.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -316,6 +348,26 @@ function (array $m) {
+    /**
+     * Disables the parsing of <style> blocks
+     *
+     * @see $isStyleBlocksParsingEnabled
```
I don't think we need this @see as the name of the field is not relevant to the external API.

oliverklee commented on README.md:
```diff
@@ -63,6 +63,18 @@ After you have set the HTML and CSS, you can call the `emogrify` method to merge
+  does not parse them, you should use this option.
+* `$emogrifier->disableInlineStylesParsing()` - By default, Emogrifier preserves all of the "style" attributes on tags
+  in the HTML you pass to it. However if you want to discard all existing inline styles in the HTML before the CSS is 
+  applied, you should use this option.
```
Could you please wrap the lines in the Markdown at 80 characters so the lines will fit for the GitHub code formatting?

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -948,6 +948,41 @@ public function emogrifyAppliesCssFromStyleNodes()
     /**
      * @test
      */
+    public function emogrifyDoesNotApplyCssFromStyleBlocksWhenDisabled()
```
Could you please move the precondition (WhenDisabled) to before the result (DoesNotApply...)?

And if you like, you can make the wording shorter by using a bit of pidgin English: DoesNotApply -> NotApplies

Ditto for the other test name.

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -948,6 +948,41 @@ public function emogrifyAppliesCssFromStyleNodes()
+        $html = $this->html5DocumentType . self::LF .
+            '<html><head><style>#topWrap p {padding-bottom: 1px;PADDING-TOP: 0;}</style></head>'
+            . '<body><div id="topWrap"><p style="text-align: center;">some content</p></div></body></html>';
+        $expected = '<p style="padding-bottom: 1px; padding-top: 0;">';
```
Currently, this test also tests that style rules are overwritten. Could you please simplify it so that it's down to the bare bones?

oliverklee commented on Tests/Unit/EmogrifierTest.php:
```diff
@@ -948,6 +948,41 @@ public function emogrifyAppliesCssFromStyleNodes()
+
+        $this->assertNotContains(
+            '<html style="' . $styleAttributeValue . '">',
+            $this->subject->emogrify()
```
Good test!

erikhansen commented:
@oliverklee Great feedback, thanks. I've implemented the requested feedback and pushed the rebased commit. 

I struggled with the names of the private methods: `saveOriginalStyle` and `normalizeOriginalStyles` are what I came up with.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -102,6 +102,26 @@ class Emogrifier
+     *
+     * @var bool
+     */
+    private $isInlineStylesParsingEnabled = true;
```
Hm, what do you think about renaming the field and the setter to contain something like "...StyleAttributes..."?

erikhansen replied:
@oliverklee I like your suggestion. Are you thinking:
`isInlineStyleAttributesParsingEnabled`
or 
`isStyleAttributesParsingEnabled`
I prefer the former, but just wanted to check.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -190,25 +210,13 @@ public function emogrify()
-                    );
-                    $this->visitedNodes[$nodePath] = $node;
+                if ($this->isInlineStylesParsingEnabled) {
+                    // Normalize the "style" attribute and save it for later retrieval
```
I don't think that we need all these comments above the method calls if the methods have good, speaking names and the method PHPDoc are good.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -397,6 +423,57 @@ public function removeUnprocessableHtmlTag($tagName)
     }
 
     /**
+     * Normalize the value of the "style" attribute and save it.
```
Third person, please: Normalizes

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -397,6 +423,57 @@ public function removeUnprocessableHtmlTag($tagName)
     /**
+     * Normalize the value of the "style" attribute and save it.
+     *
+     * @param $node
```
Please add the type for the @param and the type hinting in the method declaration.

erikhansen commented:
@oliverklee I implemented the requested changes and pushed a rebased commit. Per [this comment](https://github.com/jjriv/emogrifier/pull/156#discussion_r30236962), I opted for `isInlineStyleAttributesParsingEnabled`

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -397,6 +420,57 @@ public function removeUnprocessableHtmlTag($tagName)
+     *
+     * @return void
+     */
+    private function saveOriginalStyle(\DOMElement $node)
```
Actually, this method does not really "save" the original styles to somewhere else. It just normalizes the contents of the style attributes and overwrites them in-place. So I'd opt for naming this method "normalizeStyleAttributes".

oliverklee replied:
Or am I completely misreading the code?

erikhansen replied:
Yeah, you are correct. Your suggested name better reflects the purpose of the method. I changed this in my new commit.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -397,6 +420,57 @@ public function removeUnprocessableHtmlTag($tagName)
+        $nodePath = $node->getNodePath();
+        if (!isset($this->styleAttributesForNodes[$nodePath])) {
+            $this->styleAttributesForNodes[$nodePath] = $this->parseCssDeclarationBlock(
+                $normalizedOriginalStyle
```
Will this fit in one line (< 130 characters) now that the refactoring has allowed us to reduce the indentation level?

erikhansen replied:
Yes, it does. Good catch.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -216,7 +222,10 @@ function (array $m) {
         $allCss = $this->css;
 
-        $allCss .= $this->getCssFromAllStyleNodes($xpath);
+        // Only parse the <style> blocks if that option is enabled
```
This comment is not necessary anymore either as the code already is clear enough.

oliverklee commented on Classes/Emogrifier.php:
```diff
@@ -397,6 +420,57 @@ public function removeUnprocessableHtmlTag($tagName)
+     *
+     * return @void
+     */
+    private function normalizeOriginalStyles()
```
If I understand this (historic) code correctly, this piece of code fills in the style attributes with the styles from the attributes and the style nodes mashed together, and we should rename the method to something like "fillStyleAttributesWithMergedStyles". Or am I misreading this?

oliverklee replied:
So we should have more unit tests (a.k.a. "documentation tests") that test what functionality still remains even if the options are switched off.

erikhansen replied:
I changed the method name, as you are correct about the purpose of the method.

Can you provide some suggestions for what additional tests would test for? The emogrifyWhenDisabledNotAppliesCssFromInlineStyles test already ensures that if the "inline styles" option is disabled, inline styles get ignored. What other things should we test for?

oliverklee replied:
I'd propose two additional tests:

1. when the parsing of style nodes is disabled, that the existing style attributes are kept
2. when the parsing of style attributes is disabled, that the style nodes are still parsed and processed into the style attributes (this is where I suspect a regression)

erikhansen replied:
Great test suggestions. I've implemented the tests as `emogrifyWhenStyleBlocksParsingDisabledKeepInlineStyles` and `emogrifyWhenInlineStyleAttributesParsingDisabledKeepStyleBlockStyles` and have pushed the updated commit.

oliverklee commented:
Great work! Thanks for contributing!
